tests/explicit-nulls/warn/i21577.scala

```diff
@@ -0,0 +1,24 @@
+def f(s: String) =
+  val s2 = s.trim()
+  s2 match
+    case s3: String => println(1)
+    case _ => println(2) // warn
+
+
+def f2(s: String | Null) =
+  val s2 = s.nn.trim()
+    s2 match
+      case s3: String => println(1)
+      case _ => println(2) // warn
+
+def f3(s: String | Null) =
+  val s2 = s
+    s2 match
+      case s3: String => println(1)
+      case _ => println(2) // warn
```

Comment:
I expect no warning here.

Reply:
Isn't this case only reachable by Null? In which case we should emit a warning that it's unreachable except for null.

Reply:
Well, with explcit nullls, we can really consider `String` and `Null` as two independent types, similar to `String` and `Int`, so the semantic is clear the second case matches `null` value. What do you think? cc @olhotak

Reply:
Well, without explicit nulls, this warns. I think we can keep the warning. People can still write `case null =>` for this, which is more explicit. I think it would get tricky trying to distinguish between `String | Null` and a flexible `String` type for the scrutinee. If we start getting complaints about the warning, we can revisit it in the future.

Reply:
If the type is `String | Null` with explcit nullls, I just find it confusing to call the wildcard case "unreachable", because the types clearly indicate what values are covered or not.

Reply:
So is the intention for this warning to issue it only when the scrutinee has a flexible type?

How would we best test for that?

Reply:
I think the `isNullable` is the problem. There are some other things I don't like about this part of code. I will try to do 
 a refactor to this part.